internal/events/block_confirmations.go

```diff
@@ -0,0 +1,520 @@
+		added:         time.Now(),
+		confirmations: make([]*blockInfo, 0, bcm.requiredConfirmations),
+		event:         event,
+		eventStream:   eventStream,
+	}
+	bcm.pending[eventKey] = pending
+	bcm.log.Infof("Added pending event %s", eventKey)
+	return pending
+}
+
+// removeEvent is called by the goroutine on receipt of a remove event notification
+func (bcm *blockConfirmationManager) removeEvent(event *eventData) {
+	bcm.log.Infof("Removing stale event %s", bcm.keyForEvent(event))
+	delete(bcm.pending, bcm.keyForEvent(event))
+}
+
+func (bcm *blockConfirmationManager) processBlockHashes(blockHashes []*ethbinding.Hash) {
+	if len(blockHashes) > 0 {
+		bcm.log.Debugf("New block notifications %v", blockHashes)
+	}
+
+	for _, blockHash := range blockHashes {
+		// Get the block header
+		block, err := bcm.getBlockByHash(blockHash)
+		if err != nil || block == nil {
+			bcm.log.Errorf("Failed to retrieve block %s: %v", blockHash, err)
+			continue
+		}
+
+		// Process the block for confirmations
+		bcm.processBlock(block)
+
+		// Update the highest block (used for efficiency in chain walks)
+		if uint64(block.Number) > bcm.highestBlockSeen {
+			bcm.highestBlockSeen = uint64(block.Number)
+		}
+	}
+}
+
+func (bcm *blockConfirmationManager) processBlock(block *blockInfo) {
+
+	// Go through all the events, adding in the confirmations, and popping any out
+	// that have reached their threshold. Then drop the log before logging/processing them.
+	parentStr := block.ParentHash.String()
+	blockNumber := uint64(block.Number)
+	var confirmed pendingEvents
+	for eventKey, pending := range bcm.pending {
+		// The block might appear at any point in the confirmation list
+		expectedParentHash := pending.event.BlockHash
+		expectedBlockNumber := pending.event.blockNumber + 1
+		for i := 0; i < (len(pending.confirmations) + 1); i++ {
+			if parentStr == expectedParentHash && blockNumber == expectedBlockNumber {
+				pending.confirmations = append(pending.confirmations[0:i], block)
+				bcm.log.Infof("Confirmation %d at block %d / %s event=%s",
+					len(pending.confirmations), block.Number, block.Hash.String(), bcm.keyForEvent(pending.event))
+				break
+			}
+			if i < len(pending.confirmations) {
+				expectedParentHash = pending.confirmations[i].Hash.String()
+			}
+			expectedBlockNumber++
+		}
+		if len(pending.confirmations) >= bcm.requiredConfirmations {
+			delete(bcm.pending, eventKey)
+			confirmed = append(confirmed, pending)
+		}
+	}
+
+	// Sort the events to dispatch them in the correct order
+	sort.Sort(confirmed)
+	for _, c := range confirmed {
+		bcm.dispatchConfirmed(c)
+	}
+
+}
+
+// dispatchConfirmed drive the event stream for any events that are confirmed, and prunes the state
+func (bcm *blockConfirmationManager) dispatchConfirmed(confirmed *pendingEvent) {
+	eventKey := bcm.keyForEvent(confirmed.event)
+	bcm.log.Infof("Confirmed with %d confirmations event=%s", len(confirmed.confirmations), eventKey)
+	delete(bcm.pending, eventKey)
```

Comment:
Wouldn't we have already removed the key from `bcm.pending` since we moved this event from `bcm.pending` to `confirmed`?

Reply:
Agreed reviewing code